worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -530,30 +538,44 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
         s"While handling PushMergedData, throw $cause, fileWriter $fileWriterWithException has exception.",
         fileWriterWithException.getException)
       workerSource.incCounter(WorkerSource.WRITE_DATA_FAIL_COUNT)
-      callbackWithTimer.onFailure(new CelebornIOException(cause))
+      pushMergedDataCallback.onFailure(new CelebornIOException(cause))
       return
     }
 
-    if (fileWriters.exists(checkDiskFull(_) == true)) {
-      val (mapId, attemptId) = getMapAttempt(body)
-      logWarning(
-        s"return hard split for disk full with shuffle $shuffleKey map $mapId attempt $attemptId")
-      callbackWithTimer.onSuccess(ByteBuffer.wrap(Array[Byte](StatusCode.HARD_SPLIT.getValue)))
-      return
+    fileWriters.zipWithIndex.foreach {
+      case (fileWriter, index) =>
+        if (fileWriter == null) {
+          if (!pushMergedDataCallback.containIndex(index)) {
+            pushMergedDataCallback.onFailure(new CelebornIOException(s"Partition $index's fileWriter not found, but it hasn't been identified in the previous validation step."))
+            return
+          }
+        } else if (checkDiskFull(fileWriter)) {
+          logWarning(
+            s"return hard split for disk full with shuffle $shuffleKey map $mapId attempt $attemptId")
+          pushMergedDataCallback.addUnsuccessfulPartition(index, StatusCode.HARD_SPLIT)
+        } else if (fileWriter.isClosed) {
+          val fileInfo = fileWriter.getCurrentFileInfo
+          logWarning(
+            s"[handlePushMergedData] FileWriter is already closed! File path ${fileInfo.getFilePath} " +
+              s"length ${fileInfo.getFileLength}")
+          pushMergedDataCallback.addUnsuccessfulPartition(index, StatusCode.HARD_SPLIT)
+        } else {
+          val splitStatus = checkDiskFullAndSplit(fileWriter, isPrimary)
+          if (splitStatus == StatusCode.HARD_SPLIT) {
+            logWarning(
+              s"return hard split for disk full with shuffle $shuffleKey map $mapId attempt $attemptId")
+            workerSource.incCounter(WorkerSource.WRITE_DATA_HARD_SPLIT_COUNT)
+            pushMergedDataCallback.addUnsuccessfulPartition(index, StatusCode.HARD_SPLIT)
+          } else if (splitStatus == StatusCode.SOFT_SPLIT) {
+            pushMergedDataCallback.addUnsuccessfulPartition(index, StatusCode.SOFT_SPLIT)
+          }
+        }
+        if (!pushMergedDataCallback.containIndex(index) || pushMergedDataCallback.getStatusCode(
```

Comment:
The method getStatusCode will return -1 by default. So this check can be simplified to `pushMergedDataCallback.getStatusCode(
            index) == StatusCode.SOFT_SPLIT.getValue`

Reply:
The check could not be simplified due to some partitions that have not been split or have failed, and therefore, they should increment pending writes.